### PR TITLE
Implement foster offer flow

### PR DESCRIPTION
**Pull request by laurenmcp**
## PR Title

Issue Number(s): #104.

What does this PR change and why?

Implement foster offer questionnaire
- Implement question modal per the figma designs
- Save question responses into state
- Send email upon form submission, displaying toast messages for successful/unsuccessful requests

### Checklist

- [ ] Requirements have been implemented
- [ ] Acceptance criteria are met
- [ ] Database schema docs have been updated or are not necessary
- [ ] Code follows design and style guidelines
- [ ] Commits follow guidelines (concise, squashed, etc)
- [x] Relevant reviewers (Senior Dev/EM/Designers) have been assigned to this PR

### Critical Changes

- Database change/migration to run
- Environment config change
- Breaking API change

### Related PRs

- #number_of_pr

### Testing

Enumerate steps to test the functionality of your ticket. This should include edge cases and testing of error handling, if applicable.

### CODE REVIEW

**Comment by laurenmcp** [comment]
Notes/Todo: (aka please don't spend your time on these issues @katsukixyz!)
- refactor to include `_id` in `IPost` rather than using `IPostWithId`.
- replace hard coded email with the current user's email and (time permitting) logic to determine recipient email based on environment. 
- style toasts to match figma.
- [in progress] check with product/design on getting the full set of questions/verbiage for each foster type.

File: web/components/PetPostModal/PetPostModal.tsx [hunk 1]
```diff
@@ -25,20 +49,17 @@ type FosterTypeData = {
   }>;
 };
 
+enum QuestionKeys {
+  numOtherDogs = "numOtherDogs",
+  maxDogs = "maxDogs",
+}
```
Comment (katsukixyz) [inline web/components/PetPostModal/PetPostModal.tsx]: I think it may be unnecessary and a bit tedious to define all the question keys as an enum here - I think we can leave it as a string

File: web/components/PetPostModal/PetPostModal.tsx [hunk 2]
```diff
@@ -25,20 +49,17 @@ type FosterTypeData = {
...
+
+type QuestionResponseData = {
+  [key in QuestionKeys]?: string;
```
Comment (katsukixyz) [inline web/components/PetPostModal/PetPostModal.tsx]: Can probably just simplify to `[key: string]: string`

File: web/components/PetPostModal/PetPostModal.tsx [hunk 3]
```diff
@@ -102,13 +103,78 @@ const data: FosterTypeData = {
...
+  const initialQuestionResponseData: QuestionResponseData = data[
+    fosterType
+  ].reduce(
+    (responses, question) =>
+      ({
+        ...responses,
+        [question.key]: "",
+      } as QuestionResponseData),
+    {}
+  );
```
Comment (katsukixyz) [inline web/components/PetPostModal/PetPostModal.tsx:145]: Consider making a zod schema to define your validation like this:

```js
const formSchema = z.record(
  z.string().min(1, { message: "your error message here" })
);
```

File: web/components/PetPostModal/PetPostModal.tsx [hunk 4]
```diff
@@ -102,13 +103,78 @@ const data: FosterTypeData = {
...
+  const toast = useToast();
+
+  function handleSubmission() {
+    let formIsValid = true;
+    for (const [key] of Object.entries(fosterQuestionResponses)) {
+      if (fosterQuestionResponses[key as keyof typeof QuestionKeys] === "") {
+        formIsValid = false;
+        break;
+      }
+    }
```
Comment (katsukixyz) [inline web/components/PetPostModal/PetPostModal.tsx]: You can `safeParse()` the zod schema you create to handle the validation. See `PostCreationModal` for an existing implementation of this

File: web/components/PetPostModal/PetPostModal.tsx [hunk 5]
```diff
@@ -125,11 +191,35 @@ const FosterQuestionnaire = ({
...
-                  <Text>{question.title}</Text>
-                  <Input />
+                <Flex key={key} flexDir="column" gap={2}>
+                  <FormControl
+                    isRequired
+                    isInvalid={
+                      fosterQuestionResponses[
+                        key as keyof typeof QuestionKeys
+                      ] === ""
+                    }
```
Comment (katsukixyz) [inline web/components/PetPostModal/PetPostModal.tsx]: Let's remove this - let's only validate the responses when a user tries to submit the form.

File: web/components/PetPostModal/PetPostModal.tsx [hunk 6]
```diff
@@ -125,11 +191,35 @@ const FosterQuestionnaire = ({
...
+                      }
+                      onChange={(event) => {
+                        setfosterQuestionResponses({
+                          ...fosterQuestionResponses,
+                          [key as keyof typeof QuestionKeys]:
+                            event.target.value,
+                        });
+                      }}
+                    />
+                    <FormErrorMessage>Field required.</FormErrorMessage>
```
Comment (katsukixyz) [inline web/components/PetPostModal/PetPostModal.tsx]: See above

File: web/components/PetPostModal/PetPostModal.tsx [hunk 7]
```diff
@@ -162,11 +254,35 @@ const FosterQuestionnaire = ({
           </Stack>
           <ModalHeader>Foster Questionnaire</ModalHeader>
           <Flex paddingX={6} flexDir="column" gap={4} overflowY="scroll">
-            {data[fosterType].map((question) => {
+            {data[fosterType].map(({ key, title }) => {
               return (
-                <Flex key={question.key} flexDir="column" gap={2}>
-                  <Text>{question.title}</Text>
-                  <Input />
+                <Flex key={key} flexDir="column" gap={2}>
```
Comment (katsukixyz) [inline web/components/PetPostModal/PetPostModal.tsx]: Verify that the comments above are consistent for the mobile view as well - maybe we should revisit refactoring this code so that we don't have repeated parts like this? You can make it a TODO comment since it's outside the scope of the ticket.

**Review by katsukixyz — COMMENTED** [review]
Great start! No major concerns, just a few engineering tweaks!

File: web/components/PetPostModal/PetPostModal.tsx [hunk 8]
```diff
@@ -13,10 +15,33 @@ import {
...
+  Trained,
+  goodWithLabels,
+  genderLabels,
+  breedLabels,
+  sizeLabels,
+  ageLabels,
+  fosterTypeDescriptions,
+} from "../../utils/types/post";
+import { trpc } from "../../utils/trpc";
+import { ObjectId } from "mongoose";
```
Comment (katsukixyz) [inline web/components/PetPostModal/PetPostModal.tsx]: Use `Types.ObjectId` instead

File: web/components/PetPostModal/PetPostModal.tsx [hunk 9]
```diff
@@ -25,122 +50,232 @@ type FosterTypeData = {
...
 };
 
 const FosterQuestionnaire = ({
   fosterType,
+  postId,
   isFormViewOpen,
   onFormViewClose,
 }: {
   fosterType: FosterType;
+  postId: ObjectId;
```
Comment (katsukixyz) [inline web/components/PetPostModal/PetPostModal.tsx]: See above

File: web/utils/types/post.ts [hunk 10]
```diff
@@ -1,3 +1,5 @@
+import { ObjectId } from "mongoose";
```
Comment (katsukixyz) [inline web/utils/types/post.ts]: Use `Types.ObjectId`

File: web/utils/types/post.ts [hunk 11]
```diff
@@ -343,6 +345,9 @@ export interface IPost {
   attachments: string[];
 }
 
+export interface IFeedPost extends IPost {
+  _id: ObjectId;
```
Comment (katsukixyz) [inline web/utils/types/post.ts]: Update here

File: web/server/routers/post.ts [hunk 12]
```diff
@@ -168,7 +168,9 @@ export const postRouter = router({
       }
       try {
         const post = await getPost(input.postOid, true);
-        const email = fosterTypeEmails[post.type];
+        const email = process.env.production
```
Comment (katsukixyz) [inline web/server/routers/post.ts]: Do you think we could do `process.env.CONTEXT` instead? Take a look at https://docs.netlify.com/configure-builds/environment-variables/#build-metadata and see if that makes sense to you too

File: web/components/PetPostModal/PetPostModal.tsx [hunk 13]
```diff
@@ -25,122 +50,232 @@ type FosterTypeData = {
...
+  };
+
   return (
     <Modal
       isOpen={isFormViewOpen}
       onClose={onFormViewClose}
       isCentered
       scrollBehavior="inside"
-      size={["full", "xl"]}
+      size={["full", "l"]}
```
Comment (katsukixyz) [inline web/components/PetPostModal/PetPostModal.tsx]: Typo for `lg`?

**Review by katsukixyz — CHANGES_REQUESTED** [review]
Nice work! I'm running into an issue where I can't input more than one character at a time into the input fields without the inputs becoming unfocused - could you investigate this?

Otherwise, just some minor comments!

File: web/server/routers/post.ts (hunk 12 above)
Comment (laurenmcp) [reply web/server/routers/post.ts]: That seems reasonable to me. `process.env.NODE_ENV` is a possible alternative but I think it requires a `NODE_ENV` variable to be set (more complicated) as described here https://docs.netlify.com/configure-builds/manage-dependencies/#node-js-environment

**Comment by laurenmcp** [comment]
> Nice work! I'm running into an issue where I can't input more than one character at a time into the input fields without the inputs becoming unfocused - could you investigate this?
> 
> Otherwise, just some minor comments!

The input focus issue was due to abstracting the form input mapping/logic into its own component, so each change event forced a re-render of the entire component. I added the TODO tag to refactor so that the input mapping/logic is not duplicated for web and mobile.

File: web/components/PetPostModal/PetPostModal.tsx [hunk 14]
```diff
@@ -38,10 +38,10 @@ import {
   fosterTypeDescriptions,
 } from "../../utils/types/post";
 import { trpc } from "../../utils/trpc";
-import { ObjectId } from "mongoose";
 import { useState } from "react";
 import { z } from "zod";
 import { useAuth } from "../../context/auth";
+import { Schema } from "mongoose";
```
Comment (katsukixyz) [inline web/components/PetPostModal/PetPostModal.tsx]: Sorry, I should've been more clear! We want
```js
import { Types } from "mongoose";
Types.ObjectId
```
 This is different from `Schema.Types.ObjectId`: https://mongoosejs.com/docs/schematypes.html#what-is-a-schematype

File: web/components/PetPostModal/PetPostModal.tsx [hunk 15]
```diff
@@ -136,14 +136,17 @@ const data: FosterTypeData = {
...
+ **/
 const FosterQuestionnaire = ({
   fosterType,
   postId,
   isFormViewOpen,
   onFormViewClose,
 }: {
   fosterType: FosterType;
-  postId: ObjectId;
+  postId: Schema.Types.ObjectId;
```
Comment (katsukixyz) [inline web/components/PetPostModal/PetPostModal.tsx]: Update here

File: web/utils/types/post.ts [hunk 16]
```diff
@@ -346,7 +346,7 @@ export interface IPost {
 }
 
 export interface IFeedPost extends IPost {
-  _id: ObjectId;
+  _id: Schema.Types.ObjectId;
```
Comment (katsukixyz) [inline web/utils/types/post.ts]: Update here

File: web/server/routers/post.ts [hunk 17]
```diff
@@ -168,9 +168,10 @@ export const postRouter = router({
       }
       try {
         const post = await getPost(input.postOid, true);
-        const email = process.env.production
-          ? fosterTypeEmails[post.type]
-          : input.email;
+        const email =
+          process.env.CONTEXT === "production"
+            ? fosterTypeEmails[post.type]
+            : input.email;
```
Comment (katsukixyz) [inline web/server/routers/post.ts:174]: From my limited testing, this seems to work well!

**Review by katsukixyz — CHANGES_REQUESTED** [review]
Just type changes! Thanks for the changes Lauren!

**Review by Wesley257 — APPROVED** [review]

File: web/globalConfig.json
Comment (katsukixyz) [inline web/globalConfig.json:1]: Is this file necessary?

**Review by katsukixyz — APPROVED** [review]
